sw/host/opentitanlib/src/util/status.rs

```diff
@@ -0,0 +1,239 @@
+use std::ffi::CString;
+use std::path::PathBuf;
+
+pub use bindgen::status::absl_status_t as RawStatusCode;
+pub use bindgen::status::status_t as RawStatus;
+
+// FIXME: is there a better way of doing this? bindgen does not
+// support adding custom derive to a type.
+#[derive(Debug, serde::Serialize, serde::Deserialize, TryFromPrimitive, PartialEq, Eq)]
+#[repr(u32)]
+pub enum StatusCode {
+    Ok = bindgen::status::absl_status_code_kOk,
+    Cancelled = bindgen::status::absl_status_code_kCancelled,
+    Unknown = bindgen::status::absl_status_code_kUnknown,
+    InvalidArgument = bindgen::status::absl_status_code_kInvalidArgument,
+    DeadlineExceeded = bindgen::status::absl_status_code_kDeadlineExceeded,
+    NotFound = bindgen::status::absl_status_code_kNotFound,
+    AlreadyExists = bindgen::status::absl_status_code_kAlreadyExists,
+    PermissionDenied = bindgen::status::absl_status_code_kPermissionDenied,
+    ResourceExhausted = bindgen::status::absl_status_code_kResourceExhausted,
+    FailedPrecondition = bindgen::status::absl_status_code_kFailedPrecondition,
+    Aborted = bindgen::status::absl_status_code_kAborted,
+    OutOfRange = bindgen::status::absl_status_code_kOutOfRange,
+    Unimplemented = bindgen::status::absl_status_code_kUnimplemented,
+    Internal = bindgen::status::absl_status_code_kInternal,
+    Unavailable = bindgen::status::absl_status_code_kUnavailable,
+    DataLoss = bindgen::status::absl_status_code_kDataLoss,
+    Unauthenticated = bindgen::status::absl_status_code_kUnauthenticated,
+}
+
+// Safe abstraction over status_create
+fn status_create_safe(code: StatusCode, mod_id: u32, file: String, arg: i32) -> RawStatus {
+    // We do not expect an error since it is a valid String.
+    let file = CString::new(file).expect("CString::new failed");
+    unsafe {
+        // Safety: the function expects a valid readonly C-string which is exactly what
+        // CString:as_ptr() provides.
+        status_create(code as u32, mod_id, file.as_ptr(), arg)
+    }
+    // Note: file is dropped here so the C-string pointer is valid accross the function call.
+}
+
+// Convert an array of i8 to a string. This function will stop at first 0 (or at the
+// end of the array if it contains no zero).
+fn c_string_to_string(array: &[i8]) -> String {
+    let array = array
+        .iter()
+        .map(|c| *c as u8)
+        .take_while(|c| *c != 0u8)
+        .collect::<Vec<_>>();
+    String::from_utf8_lossy(&array).to_string()
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+/// Hold a completely decoded status.
+pub struct Status {
+    // Module ID if present.
+    pub module_id: String,
+    // Status argument. Note that depending on whether this is an error status or ok
+    // status, the maximum value that can fit is different.
+    pub arg: i32,
+    // Error code of the status.
+    pub code: StatusCode,
+}
+
+// Safe abstraction over status_extract.
+impl TryFrom<RawStatus> for Status {
```

Comment:
`TryFrom`/`From` feels a bit weird in this case because usually they have conversion semantics, not parsing semantics. I'd recommend just have inherent functions like `from_raw_status` and `from_u32`.

Reply:
There is no parsing involved, this is just converting between a packed representation using bit fields to a more usable rust structure.
Can you elaborate on what you mean by the distinction between `From` and `from_*`? If there was parsing involved, I would rather call it `parse_u32` or `parse_from_u32`.

Reply:
Generally if `From/TryFrom` don't do conversions that may have special semantics, e.g. `from_utf8` is a function but there is no `TryFrom<&[u8]> for &str`. But the line is pretty vague.